_posts/2021-04-22-slow-shader.md

```diff
@@ -0,0 +1,132 @@
+
+I have a [blog post planned about clipping] in piet-gpu, but it probably helps to explain a little what's going on. An arbitrary 2D scene is potentially a tree of nested clips. Each clip node in the tree has an associated clip mask. When rendering the clip masks with antialiasing, one way to describe the desired result is to render the children of the clip node into an RGBA buffer, render the clip mask itself into an alpha buffer, and then alpha-composite that (multiplying the two alpha values together) to the render target surface.
+
+The way this works in piet-gpu is that the earlier stages of the pipeline (coarse rasterization) prepare a per-tile command list, which is then interpreted in the last stage (fine rasterization). Commands include `BeginClip` and `EndClip`, which push and pop these temporary pixel values onto a stack. Because the nesting depth of clips is not limited, this stack can be of arbitrary depth. We've spent some time exploring hybrid approaches where there's a small window in registers and it spills to memory when that window is exceeded, but at present it always goes to memory.
+
+## Reading shader assembly
+
+At this point, I had no desire to remain an ignoramus; I felt that we must know what's really going on, and determined that we shall know. The next step was to [read shader assembly][How To Read Shader Assembly].
+
+For certain GPUs, especially Radeon, this is easy, as the shader compiler is open source and widely available. In fact, the Radeon shader analyzer is available on the [Shader playground], a resource analogous to [Godbolt] for shaders. But for mobile GPUs, it's a bit harder, as the drivers are proprietary.
+
+Fortunately, for someone as determined as I was, it's possible, largely thanks to the existence of free tools, especially [Freedreno]. I was prepared to go as deep as I needed, including getting Freedreno working on some hardware; if it performed well, that would point to the proprietary driver. But if the performance matched, I would easily be able to see the ISA it produced, and dig into that.
+
+As it turned out, I didn't have to go that far. The [glGetProgramBinary] function can be used to spit out a binary, and then the [disassembly tools] in Mesa could spit out disassembler for me to read. There's a bit more complexity to this, as the drivers produce both a compressed and an uncompressed format. The corresponding Vulkan call is [vkGetPipelineCacheData], but I was only able to get the compressed form. I strongly suspect it's possible to work out the compression format, but when I got readable results from the GL route, that was good enough for me.
+
+Looking at the disassembly (the full traces are in the [piet-gpu bug]), one thing immediately stuck out. In the happy path, all reads from memory use the `isam` (Image SAMple) instruction, like this:
+
+```
+:5:0166:0205[a0005100x_02000001x] isam (s32)(x)r0.x, r0.x, s#0, t#1
```

Comment:
Suggest dropping stuff before `isam` to reduce unnecessary clutter. Seam with the `ldib` instruction below.